src/main/java/com/atlassian/migration/datacenter/core/application/JiraConfiguration.java

```
@@ -0,0 +1,90 @@
+import org.springframework.stereotype.Component;
+import org.xml.sax.InputSource;
+
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathFactory;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+
+/**
+ * Copyright Atlassian: 04/03/2020
+ */
+@Component
+public class JiraConfiguration implements ApplicationConfiguration
+{
+    private JiraHome jiraHome;
+    private Optional<DatabaseConfiguration> databaseConfiguration = Optional.empty();
+
+    public JiraConfiguration(JiraHome jiraHome)
+    {
+        this.jiraHome = jiraHome;
+    }
+
+    private String getXPath(String path, Path config) throws ConfigurationReadException
+    {
+        // We need to recreate the input stream for each read; not the most efficient,
+        // but it's only done once and it's simpler than generating a full DOM.
+        FileReader reader;
+        try {
+            reader = new FileReader(config.toFile());
+        } catch (FileNotFoundException e) {
+            throw new ConfigurationReadException("Failed to open dbconfig.xml", e);
+        }
+
+        InputSource inputSource = new InputSource(reader);
+        XPath xPath = XPathFactory.newInstance().newXPath();
+        try {
+
+            String val = xPath.evaluate(path, inputSource);
+            // Note: XPath will return an empty string if not found;
+            // we assume retrieved values can never be empty.
+            if (val == null || val.equals(""))
+                throw new ConfigurationReadException("No value found for XPath: "+path);
+
+            return val;
+
+        } catch (Exception e) {
+            throw new ConfigurationReadException("Failed to parse dbconfig.xml, XPath: "+path, e);
+        }
+    }
+
+    private DatabaseConfiguration parseDbConfig() throws ConfigurationReadException
+    {
```

I believe the password can be Base64 "encrypted" in the DB config.xml. There is a property in the XML which is set if this is true. We probably need to handle that case.

On further inspection, it looks like there are multiple supported methods for [encrypting the database password](https://confluence.atlassian.com/adminjiraserver/encrypting-database-password-974378811.html). It may be too difficult to decrypt the database password so perhaps we should throw an exception if we see a cipher element in the `dbconfig.xml` and give them the chance to put the password in via UI?

---

Useful info, thank. For MVP I'll implement the basic ones and throw an exception for the rest, we can add something more advanced after that.